src/audio/module_adapter/library/native_system_service.c

```diff
@@ -128,12 +128,21 @@ AdspErrorCode native_system_service_get_interface(adsp_iface_id id, system_servi
 	return ADSP_NO_ERROR;
 }
 
+#include <sof/audio/component.h>
+#include <sof/audio/data_blob.h>
+#include <sof/math/numbers.h>
+
 struct native_system_service_api native_system_service = {
 	.log_message  = native_system_service_log_message,
 	.safe_memcpy = native_system_service_safe_memcpy,
 	.safe_memmove = native_system_service_safe_memmove,
 	.vec_memset = native_system_service_vec_memset,
 	.notification_create = native_system_service_create_notification,
 	.notification_send = native_system_service_send_notif_msg,
-	.get_interface = native_system_service_get_interface
+	.get_interface = native_system_service_get_interface,
+	.comp_verify_params = comp_verify_params,
+	.math_gcd = gcd,
+	.data_blob_handler_new = comp_data_blob_handler_new_ext,
+	.data_blob_handler_free = comp_data_blob_handler_free,
+	.data_blob_set = comp_data_blob_set,
```

Comment:
Please remove unused services, there is no call of them in module.

Reply:
it is used in smart_amp module
````c
static int smart_amp_init(struct processing_module *mod)
{
        ......
	/* component model data handler */
	sad->model_handler = mod->sys_service->data_blob_handler_new(dev);
	if (!sad->model_handler) {
		ret = -ENOMEM;
		goto sad_fail;
	}
    .....
}
````

Reply:
This way we are mixing module api with component api. System services are module api only.

Reply:
@lyakh, We must discuss in wider audience how system services work and how to make smart amp work in such a narrow environment. System services just expose some interfaces and we cannot expand them anytime when we need it. Plus mixing compnen api with module api is no go.

Reply:
i'm not sure if this is a case, but in IPC4 blobs should be passed by large_config_set. In case of huge blobs several conf_sets should be sent and re-assembled in the module

Reply:
> This way we are mixing module api with component api. System services are module api only.

@ranj063 Can you please comment here? Most FW processing modules need to handle data blobs. Is there a module API function for this need?

Reply:
> i'm not sure if this is a case, but in IPC4 blobs should be passed by large_config_set. In case of huge blobs several conf_sets should be sent and re-assembled in the module

@marcinszkudlinski @lgirdwood @ranj063 
Module API provides large_config_set API for host to pass blobs to the module. Now the questions is that how the module store and convert the huge blobs via several conf_sets to its internal configuration data blob or structure. Built-in modules can leverage blob handler provided by the base FW. But for loadable modules, is there a general way to do so? If there is no general implementation solution, each loadable module needs to implement by itself.

Reply:
blob function is a utility and used for module config saving in module initialization function here, so it is not related to large_config_set.

Reply:
data_blob.c provides a library for receiving and sending of large component (or module) configuration sets as well as for run-time querying of that configuration data by component or module drivers. It belongs to the "audio component driver" infrastructure but currently it is used by module adapter drivers too. Its dependency on the component driver API is its use of a `.dev` pointer in `struct comp_data_blob_handler`. That pointer is used for 3 purposes:
1. logging in calls to `comp_dbg(dev,...)`, `comp_warn(dev,...)`, etc.
2. component status checks via `dev->state`
3. calls to `struct comp_data_blob_handler::validator(dev,...)`

The third item is the easiest to eliminate - `.validator` is a member of `struct comp_data_blob_handler` and its prototype can easily be changed to
```
int (*validator)(struct comp_data_blob_handler *blob_handler, void *new_data, uint32_t new_data_size);
```
Items (1) and (2) so far cannot be (easily) eliminated, but they aren't different from how the current module adapter implementation is related to `struct comp_dev` - currently `struct processing_module` also contains a pointer to `struct comp_dev` and it uses it for similar purposes.
So, I don't see why the data blob functions cannot be added to the "native system services" set. That set anyway will include not only the module adapter API native methods, but also mathematics (`gcd()`), memory allocation, logging, etc. It might be true that we can modify the smart amp test module to not use the data blob API (and not include the header), then we don't need to add these functions now and we can think about that when we migrate the first module, that does need to use them.

Reply:
How many components/modules are using data_blob.c ?

Reply:
> How many components/modules are using data_blob.c ?

```
sof$ git grep "data_blob\.h" src/audio/
src/audio/crossover/crossover.c:#include <sof/audio/data_blob.h>
src/audio/data_blob.c:#include <sof/audio/data_blob.h>
src/audio/dcblock/dcblock.c:#include <sof/audio/data_blob.h>
src/audio/drc/drc.c:#include <sof/audio/data_blob.h>
src/audio/eq_fir/eq_fir.c:#include <sof/audio/data_blob.h>
src/audio/eq_iir/eq_iir.c:#include <sof/audio/data_blob.h>
src/audio/google/google_hotword_detect.c:#include <sof/audio/data_blob.h>
src/audio/google/google_rtc_audio_processing.c:#include <sof/audio/data_blob.h>
src/audio/igo_nr/igo_nr.c:#include <sof/audio/data_blob.h>
src/audio/mfcc/mfcc.c:#include <sof/audio/data_blob.h>
src/audio/mux/mux.c:#include <sof/audio/data_blob.h>
src/audio/rtnr/rtnr.c:#include <sof/audio/data_blob.h>
src/audio/tdfb/tdfb.c:#include <sof/audio/data_blob.h>
```

Reply:
We must discuss it in bigger group because if I understand correctly blob isn't generic thing and can be different per module. That is why in reference fw, blobs are internal module thing and each module is doing with blobs whatever they want. I'm not against just I need to be sure what extending native_system_services will cause in near future.

Reply:
I'm no expert at blob handling but as I understand correctly, Can FW process blob and pass it as a pointer by native system service? For me it seams unnecessary to pass all data_blob.c.

Reply:
@pjdobrowolski that's the current data-blob API, it probably can be improved and made slimmer, but this is what we currently have